manifests/init.pp

```diff
@@ -0,0 +1,188 @@
+    fail('Module etcd only supported on systems using systemd')
+  }
+  if ! $config['data-dir'] {
+    fail('Module etcd requires data-dir be specified in config Hash')
+  }
+
+  case $arch {
+    'x86_64', 'amd64': { $real_arch = 'amd64' }
+    'aarch64':         { $real_arch = 'arm64' }
+    default:           { $real_arch = $arch }
+  }
+
+  $_download_url = pick($download_url, "${base_url}/v${version}/etcd-v${version}-${os}-${real_arch}.tar.gz")
+  $install_dir = "${extract_dir}/etcd-${version}"
+
+  file { $install_dir:
+    ensure => 'directory',
+    owner  => 'root',
+    group  => 'root',
+    mode   => '0755',
+  }
+
+  archive { "${download_dir}/etcd.tar.gz":
+    source          => $_download_url,
+    extract         => true,
+    extract_path    => $install_dir,
+    extract_command => 'tar xfz %s --strip-components=1',
+    creates         => "${install_dir}/etcd",
+    cleanup         => true,
+    user            => 'root',
+    group           => 'root',
+    require         => File[$install_dir],
+    before          => [
+      File["${bin_dir}/etcd"],
+      File["${bin_dir}/etcdctl"],
+    ]
+  }
+
+  file { 'etcd':
+    ensure => 'link',
+    path   => "${bin_dir}/etcd",
+    target => "${install_dir}/etcd",
+    notify => Service['etcd'],
+  }
+  file { 'etcdctl':
+    ensure => 'link',
+    path   => "${bin_dir}/etcdctl",
+    target => "${install_dir}/etcdctl",
+  }
+
+  if $manage_user {
+    user { 'etcd':
+      ensure     => 'present',
+      name       => $user,
+      forcelocal => true,
+      shell      => '/bin/false',
+      gid        => $group,
+      uid        => $user_uid,
+      home       => $config['data-dir'],
+      managehome => false,
+      system     => true,
+      before     => Service['etcd'],
+    }
+  }
+  if $manage_group {
+    group { 'etcd':
+      ensure     => 'present',
+      name       => $group,
+      forcelocal => true,
+      gid        => $group_gid,
+      system     => true,
+      before     => Service['etcd'],
+    }
+  }
+
+  file { 'etcd.yaml':
+    ensure  => 'file',
+    path    => $config_path,
+    owner   => $user,
+    group   => $group,
+    mode    => '0600',
+    content => to_yaml($config),
+    notify  => Service['etcd'],
+  }
+
+  file { 'etcd-data-dir':
+    ensure => 'directory',
+    path   => $config['data-dir'],
+    owner  => $user,
+    group  => $group,
+    mode   => '0700',
+    before => Service['etcd'],
+  }
+
+  if $config['wal-dir'] {
+    file { 'etcd-wal-dir':
+      ensure => 'directory',
+      path   => $config['wal-dir'],
+      owner  => $user,
+      group  => $group,
+      mode   => '0700',
+      before => Service['etcd'],
+    }
+  }
+
+  systemd::unit_file { 'etcd.service':
+    content => template('etcd/etcd.service.erb'),
+    notify  => Service['etcd'],
+  }
+
+  if versioncmp($facts['puppetversion'],'6.1.0') < 0 {
+    # Puppet 5 doesn't have https://tickets.puppetlabs.com/browse/PUP-3483
```

Comment:
Puppet 5 doesn't have what?

Reply:
Puppet before 6.1.0 doesn't do `systemctl daemon-reload` when units change.

Reply:
Without that command unit file changes won't be visible to systemd.

Reply:
I'm saying the comment does not read well. Change it to something like 'Puppet 5 has a bug with systemd that requires this extra ordering' and give the link.

Reply:
Updated comment, it's not really a bug, just a lacking feature that was added in Puppet 6.1.0.

Reply:
Also the systemd module would be managing this relationship if we had that module manage the service, but we don't so we create the relationship in this module, a comment for that is also there.